src/server/transport/pubsub/generic-pub-sub.ts

```diff
@@ -0,0 +1,23 @@
+import type { Observable } from 'rxjs';
+
+/** Generic interface for pub-subs.  */
+export interface GenericPubSub<T> {
+  // Publish an event for a match.
+  publish(channelId: PubSubChannelId, payload: T);
+
+  // Subscribes to events related to a single match.
+  subscribe(channelId: PubSubChannelId): Observable<T>;
```

Comment:
I’m not 100% sure about using `rxjs` for this. I’ve had a look at their docs and it seems to add quite a bit of conceptual overhead for functionality most of which we don’t need. Insisting on using it here might make contributing additional pubsub adapters more difficult than necessary for people not familiar with `rxjs`

At the very least, can we make the use of `rxjs` an implementation detail here?

We only call `subscribe` in one place and it is used as `subscribe(id).subscribe(callback)`, so it seems we could make this more flexible, by just defining:

```suggestion
  subscribe(channelId: PubSubChannelId, callback: (payload: T) => void): void;
```

A pubsub implementation is then free to use RxJS internally if it likes, but it isn’t a requirement. (I’d lean towards removing at least for the in-memory implementation to avoid adding dependencies, but we can do whatever with the separate redis package.)

Reply:
(On a side note regarding dependencies, we don’t currently have `rxjs` in our dependencies, so that fact this works is probably just by chance because some other package is pulling it in.)

Reply:
It is very easy to get to callback hell using them, and make code that is very hard to understand. We cant use promises here because there will be multiple . RxJS use case is exactly this (streams) and gives powerful functional operations like map, filter, etc... But sure, followed your advice.